gunrock/app/rw/rw_enactor.cuh

```diff
@@ -69,50 +69,91 @@ struct RWIterationLoop : public IterationLoopBase
-          walks[write_idx] = v[i];                          // record current position in walk
-          
-          if(iteration < walk_length - 1) {
-            // Determine next neighbor to walk to
-            SizeT num_neighbors = graph.GetNeighborListLength(v[i]);
-            SizeT offset        = (SizeT)round(0.5 + num_neighbors * rand[i]) - 1;
-            VertexT neighbor    = graph.GetEdgeDest(graph.GetNeighborListOffset(v[i]) + offset);
-            v[i]                = neighbor; // Replace vertex w/ neighbor in queue            
-          }
-          
-        }, frontier.queue_length, util::DEVICE, oprtr_parameters.stream));          
+
+            // printf("graph.node_values[i]=%d\n", graph.node_values[i]);
+
+            SizeT write_idx  = (i * walk_length) + iteration; // Write location in RW array
+            walks[write_idx] = v[i];                          // record current position in walk
+
+            if(iteration < walk_length - 1) {
+              // Determine next neighbor to walk to
+              SizeT num_neighbors = graph.GetNeighborListLength(v[i]);
+              SizeT offset        = (SizeT)round(0.5 + num_neighbors * rand[i]) - 1;
+              VertexT neighbor    = graph.GetEdgeDest(graph.GetNeighborListOffset(v[i]) + offset);
+              v[i]                = neighbor; // Replace vertex w/ neighbor in queue
+            }
+          };
+
+          GUARD_CU(frontier.V_Q()->ForAll(
+            rw_op, frontier.queue_length, util::DEVICE, oprtr_parameters.stream));
+
+        } else if (walk_mode == 1) {
+          auto rw_op = [
+              graph,
+              walks,
+              rand,
+              iteration,
+              walk_length
+          ] __host__ __device__ (VertexT *v, const SizeT &i) {
+
+            SizeT write_idx  = (i * walk_length) + iteration; // Write location in RW array
+            walks[write_idx] = v[i];                          // record current position in walk
+
+            if(iteration < walk_length - 1) {
+              // Walk to neighbor w/ maximum node value
+              SizeT num_neighbors        = graph.GetNeighborListLength(v[i]);
+              SizeT neighbor_list_offset = graph.GetNeighborListOffset(v[i]);
+
+              VertexT max_neighbor_id  = graph.GetEdgeDest(neighbor_list_offset + 0);
+              VertexT max_neighbor_val = graph.node_values[max_neighbor_id];
+              for(SizeT offset = 1; offset < num_neighbors; offset++) {
+                VertexT neighbor     = graph.GetEdgeDest(neighbor_list_offset + offset);
+                ValueT  neighbor_val = graph.node_values[neighbor];
+                if(neighbor_val > max_neighbor_val) {
+                  max_neighbor_id  = neighbor;
+                  max_neighbor_val = neighbor_val;
+                }
+              }
+              v[i] = max_neighbor_id; // Replace vertex w/ neighbor in queue
+            }
+          };
```

Comment:
Just a note here, is it better to have two separate lambdas for the different random walk modes, or have the if statement just inside one single lambda?

-- on a performance note, the predictor will learn very quickly that the walk mode is constant and the if statement won't diverge.

Let me know your thoughts on this (what's the best programming practice here), if statement inside lambda to select the mode, or outside?

Reply:
I would prefer using two separate lambdas, as it is right now -- that way, someone could reasonably write a new lambda without having to touch the existing ones.